pelican/contents.py

```diff
@@ -262,22 +263,34 @@ def _link_replacer(self, siteurl, m):
                     os.path.join(self.relative_dir, path)
                 )
 
-            if path not in self._context['filenames']:
-                unquoted_path = path.replace('%20', ' ')
-
-                if unquoted_path in self._context['filenames']:
-                    path = unquoted_path
-
-            linked_content = self._context['filenames'].get(path)
+            key = '{}_content'.format('static' if what in ('static', 'attach')
+                                      else 'generated')
+
+            def _get_linked_content(key, path):
+                try:
+                    return self._context[key][path]
+                except KeyError:
+                    try:
+                        # Markdown escapes spaces, try unescaping
+                        return self._context[key][path.replace('%20', ' ')]
+                    except KeyError:
+                        if what == 'filename' and key == 'generated_content':
+                            key = 'static_content'
+                            linked_content = _get_linked_content(key, path)
+                            if linked_content:
+                                logger.warning(
+                                    'Found {}. For static content,'
+                                    '{{filename}} has been deprecated in'
+                                    'favour of {{static}}, and may be removed '
+                                    'in the future'.format(path)
+                                )
+                                return linked_content
+                        return None
+
+            linked_content = _get_linked_content(key, path)
             if linked_content:
                 if what == 'attach':
-                    if isinstance(linked_content, Static):
-                        linked_content.attach_to(self)
-                    else:
-                        logger.warning(
-                            "%s used {attach} link syntax on a "
-                            "non-static file. Use {filename} instead.",
-                            self.get_relative_source_path())
```

Comment:
So... What happened to this warning?

If my understanding is correct, you get a `Replacement indicator 'attach' not recognized` instead (which is very confusing).

Reply:
Why would you expect to see that? I can only find `Replacement indicator %s not recognized` in the `else` branch starting in line 310, but at this point we've already entered the `if` branch.

I think I removed the warning because `isinstance(linked_content, Static)` will always be true: if you use `{attach}` on e.g. the source file of an article then that source file will be attached as a static file.

Reply:
Hmm... That is tricky. 

`STATIC_EXCLUDE_SOURCES` is now useless then? At least for the auto-magically added static files? Might need to check for that.

I was assuming based on that. If you `{attach}` a source file and that file is not in `static_content`, or at least it's not supposed to be, then you get `None` for `linked_content` and fall through to the else branch in 310.

Reply:
The idea is if you're linking to a source file with `{static}`, that's an explicit confirmation that you do want that file. So `STATIC_EXCLUDE_SOURCES` only affects `STATIC_PATHS`. You might want to set it to false if you want to publish source files from an entire directory without explicitly linking to them (or in the template rather than in the input files), although I guess there is also `OUTPUT_SOURCES` so you may be right that it's not super necessary anymore.

Reply:
OK, I'm fine with the fact that explicit `{static}` linking means you *want* the source. `{attach}` is sketchy though. Maybe check if `path` is in `generated_content` for `{attach}` and issue a warning?

Reply:
Do you think people might think they can change the path where a generated file ends up through `{attach}`, and thus mistakenly use it for that purpose?

I was thinking that when people want to include a link to a source file, they're about as likely as not to want that file to sit in the same folder as the file they're linking from, and thus choose `{attach}` over `{static}`. Then they would get this warning which they can't do anything about.

Moreover, I think the documentation is pretty clear about the fact that `{attach}` only works for static content.

Reply:
I was thinking more about people accidentally using `{attach}` in place of `{filename}`, but you're right. There is a legitimate reason to use `{attach}` on a source file. I guess people need to be careful now :).